Traversal: skip folder if size < minsize

aowenson-imm opened the pull request:
Implements #662.

```
/* skip folder if size < minsize
* In some filesystems e.g. Ceph, folder size = 
* recursive sum of all contents. */
```

Massive performance improvement for me (Ceph filesystem, skipping folders < 1 GB), been using occasionally for several weeks now. If your filesystem does not do recursive folder sizing e.g. ext4, then no effect. Not sure what other filesystems implement this other than Ceph.

vassilit committed:
Merge branch 'master' into develop

using develop's new features and architecture while keeping master's
fixes and user-facing behaviour.

- `--is-reflink` exit code 5 means MAYBE_REFLINK in master, and
  INLINE_EXTENTS in develop. MAYBE_REFLINK stays 5 and INLINE_EXTENTS
  become 12. In the man page, take develop's phrasing for exit code 6
  as we got it wrong in 01f5aadd (#636). Sorry.
  Also, exit code 1 is RM_LINK_ERROR in develop but RM_LINK_NONE
  in master and vice versa. I kept master's to avoid breaking user
  scripts. Note that master returns EXIT_FAILURE(1) in some cases.

- kept `--write-unfinished` (as people using the master branch seem
  to like its speed), but as it can produce surprising results,
  its short option `-U` is changed to the the quasi-alternative
  `--hash-unmatched`. Updated tests to cover both. Added a note in man.

- kept `--without-compile-glib-schemas` over `--without-schemas-compile`.

- adapted several fixes from master to the new develop codebase

- updated gettext strings and po files

aowenson-imm committed:
Traversal: skip folder if size < minsize

aowenson-imm committed:
Traversal: skip folder if size < minsize

aowenson-imm commented:
Test fail:
```
        *_, footer = run_rmlint('--size 1024')
>       assert footer['duplicates'] == 4
E       assert 0 == 4
```

Cause is my folder-skip logic on "non-recursive-size" filesystems e.g. ext4 is skipping all folders (because size < 1KB).

Two possible solutions:
1. auto-detect if filesystem implements recursive-size for folders (does folder size exceed its metadata size)
2. new CLI argument to request applying size-skipping to folders

Which do you prefer?

vassilit commented:
rmlint already do detect filesystems for some options (support of ioctls as FICLONERANGE, etc). 

However, I think that the st_size value of CephFS with the rbytes feature enabled is non-standard, so I would prefer a new CLI long option.

aowenson-imm committed:
Enable dir-skipping via --size

aowenson-imm committed:
Enable dir-skipping via --size

aowenson-imm commented:
How about this syntax instead of a whole new argument:
> rmlint -s 1G,d

I.e. just append `,d` for directories.

aowenson-imm commented:
Hi @vassilit , is that argument syntax ok? I'd like to get this merged as I have another useful feature that overlaps slightly with this.

vassilit committed:
duplicate dirs candidates: don't count them as suspicious and fix fdupes formatter

- fdupes now ignore RM_LINT_TYPE_DUPE_DIR_CANDIDATE as other formatters do
- don't count them as suspicious items
- use an explictly named alias to use as a boundary marker
- add a test

vassilit committed:
tests: fix pedantic runs when tests uses with_json=False

It may returns an empty (non-None) list.

vassilit commented:
Hello @aowenson-imm,

You should have rebased the dev branch, not merged.
Also, it would be great to have two tests: probably not a real CephFS, but on tmpfs, assuming 4096, with a commentary.

After we finish cleaning up develop, this can be merged.

Thank you and sorry for the long wait.

aowenson-imm committed:
Port-in FLAG_NOT_TRAVERSED

aowenson-imm committed:
Add unit test

aowenson-imm commented:
rebased and test added

aowenson-imm commented:
Did you intend to close this as well as dev? Did you mean to move this to develop?

vassilit commented:
Oh, sorry, I've never closed this. Github might have hallucinated.
Or took an absurd decision in my name when I deleted the dev branch, as it's merged into develop.
develop is the official protected development branch and PR that aren't hotfixes should target it.
Please re-open the PR with a develop target.

I'm sorry again. A few rework on develop and this could be merged. It will be before release anyway.

vassilit commented:
So, this is silly. I had to recreate a dev branch, change the base here and delete dev again.

aowenson-imm commented:
Can you rebase to develop branch for me? I get merge conflicts in unrelated test files e.g.
> CONFLICT (content): Merge conflict in tests/test_options/test_size.py

Or do I rebase back to main then to develop?

vassilit commented:
I can rebase for you before review, yes.

RayOei commented:
@aowenson-imm @vassilit What is the status of this one?

aowenson-imm commented:
Waiting on @vassilit to fix the base-branch situation. Feature works perfect.

vassilit commented:
> @aowenson-imm @vassilit What is the status of this one?

I will rebase the branch before the next release, and we will benefit from the new develop that provides more testing overall.
A local Ceph FS with cephadm might be even added to the tests.

aowenson-imm commented:
FYI, that other overlapping feature I mentioned is a timeout on folder listing, I can prepare PR once this merged. But it's a bigger change than this - involves putting traversal in a separate process. I've run this for ~6 months no problem.

vassilit commented on lib/traverse.c:
```diff
@@ -403,6 +403,23 @@ static gint rm_traverse_directory(RmTravBuffer *buffer, RmSession *session) {
         } else {
             switch(p->fts_info) {
             case FTS_D: /* preorder directory */
+                if(cfg->limits_specified && cfg->apply_limits_to_dirs && cfg->minsize > 0) {
```
I've rewritten a bit here, removed unecessary casts and braces.

vassilit commented on tests/test_options/test_size.py:
```diff
@@ -102,3 +105,29 @@ def test_replay_size():
            ["emptydir"] + (["emptyfile"] * 2) + (["duplicate_file"] * 4)
+
+
+def test_directory_size_limit_on_tmpfs():
```
Test rewritten a bit to use the testsuite tools.

vassilit commented on lib/cmdline.c:
```diff
@@ -157,12 +157,26 @@ static gboolean rm_cmd_parse_limit_sizes(_UNUSED const char *option_name,
+    const char *modifier = range_parts[1];
+
+    if(modifier != NULL) {
+        if(g_ascii_strcasecmp(modifier, "d") == 0) {
```
,d
So that we could support other flags in the future.
